src/Sarif.Viewer.VisualStudio/Tags/SarifLocationErrorTagger.cs

```diff
@@ -0,0 +1,139 @@
+        private readonly ITextBuffer textBuffer;
+
+        private List<ISarifLocationTag> currentTags;
+        private bool tagsDirty = true;
+
+        /// <inheritdoc/>
+        public event EventHandler<SnapshotSpanEventArgs> TagsChanged;
+
+        /// <inheritdoc/>
+        public event EventHandler Disposed;
+
+        public SarifLocationErrorTagger(ITextBuffer textBuffer, IPersistentSpanFactory persistentSpanFactory, ISarifErrorListEventSelectionService sarifErrorListEventSelectionService)
+        {
+            ThreadHelper.ThrowIfNotOnUIThread();
+
+            if (!SdkUIUtilities.TryGetFileNameFromTextBuffer(textBuffer, out this.filePath))
+            {
+                throw new ArgumentException("Always expect to be able to get file name from text buffer.", nameof(textBuffer));
+            }
+
+            this.textBuffer = textBuffer;
+            this.persistentSpanFactory = persistentSpanFactory;
+            this.sarifErrorListEventSelectionService = sarifErrorListEventSelectionService;
+            this.sarifErrorListEventSelectionService.SelectedItemChanged += SarifErrorListEventSelectionService_SelectedItemChanged;
+        }
+
+        /// <inheritdoc/>
+        public IEnumerable<ITagSpan<IErrorTag>> GetTags(NormalizedSnapshotSpanCollection spans)
+        {
+            if (this.tagsDirty)
+            {
+                this.tagsDirty = false;
+
+                // This query is saying, create the error tags for the top-level SARIF result (includeChildTags: false, includeResultTag: true)
+                // and then create the tags for the children (includeChildTags: true, includeResultTag: false) but then
+                // filter the children (thread-flows, call nodes, etc.) to the ones relevant to the currently selected error item.
+                // This is done so that duplicate locations from multiple results in results are rendered in the editor.
+                IEnumerable<ISarifLocationTag> possibleTags = CodeAnalysisResultManager.
+                    Instance.
+                    RunIndexToRunDataCache.
+                    Values.
+                    SelectMany(runDataCache => runDataCache.SarifErrors).
+                    Where(sarifListItem => string.Compare(this.filePath, sarifListItem.FileName, StringComparison.OrdinalIgnoreCase) == 0).
+                    SelectMany(sarifListItem => 
+                    sarifListItem.GetTags<IErrorTag>(this.textBuffer, this.persistentSpanFactory, includeChildTags: false, includeResultTag: true).Concat(
+                        sarifListItem.GetTags<IErrorTag>(this.textBuffer, this.persistentSpanFactory, includeChildTags: true, includeResultTag: false).Where(
+                            sarifLocationTag => this.sarifErrorListEventSelectionService.SelectedItem != null && sarifLocationTag.ResultId == this.sarifErrorListEventSelectionService.SelectedItem.ResultId))).
```

Comment:
>this.sarifErrorListEventSelectionService.SelectedItem != null [](start = 48, length = 61)

The question of whether there's a selected item only has to be answered once. IMO this is a time where, although it's _possible_ to write this in a single LINQ query, it makes the code too hard to understand.

Reply:
I'll scrape together a proposal for a helper method `GetRelevantTags` that encapsulates this query but breaks it apart in a way I'll find easier to understand.

---
In reply to: [502676525](https://github.com/microsoft/sarif-visualstudio-extension/pull/214#discussion_r502676525) [](ancestors = 502676525)